File: 2020/09/solution.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func readInput() []int {
	input, err := ioutil.ReadFile("2020/09/input.txt")
	if err != nil {
		panic(err)
	}

	reader := strings.NewReader(string(input))
	scanner := bufio.NewScanner(reader)

	var result []int
	for scanner.Scan() {
		line := scanner.Text()
		num, _ := strconv.Atoi(line)
		result = append(result, num)
	}
	return result
}

func partOne(numbers []int) int {
	length := len(numbers)

	for i := 25; i < length; i++ {
		var sums []int
		for j := i - 25; j < i-1; j++ {
			for k := j + 1; k < i; k++ {
				sums = append(sums, numbers[j]+numbers[k])
			}
		}
		valid := false
		for _, sum := range sums {
			if numbers[i] == sum {
				valid = true
				break
			}
		}
		if !valid {
			return numbers[i]
		}
	}

	return -1
}

func partTwo(numbers []int, target int) int {
	i := 0
	j := 1
	sum := numbers[i] + numbers[j]
	for sum != target {
		if sum < target {
			j++
			sum = sum + numbers[j]
		} else {
			sum = sum - numbers[i]
			i++
		}
	}
	min := numbers[i]
	max := numbers[j]
	for k := i; k <= j; k++ {
		number := numbers[k]
		if number < min {
			min = number
		}
		if number > max {
			max = number
		}
	}
	return min + max
}

func main() {
	numbers := readInput()

	part_one := partOne(numbers)
	part_two := partTwo(numbers, part_one)

	fmt.Printf("%d\n", part_one)
	fmt.Printf("%d\n", part_two)
}

File: 2020/09/solution.exs
IO.write("\e[H\e[2J")
IO.puts("Running #{__ENV__.file}...")

numbers = __DIR__ <> "/" <> "input.txt"
  |> File.read!()
  |> String.split("\n", trim: true)
  |> Enum.map(fn line ->
    {integer, _leftover} = Integer.parse(line)
    integer
  end)

defmodule CombinationHelper do
  def combinations(list, num)
  def combinations(_, 0), do: [[]]
  def combinations([], _), do: []
  def combinations([head | tail], num) do
    Enum.map(combinations(tail, num - 1), &[head | &1]) ++
      combinations(tail, num)
  end
end

defmodule SumBuilder do
  def build(list, index)
  def build(_, index) when index < 25, do: false
  def build(list, index) do
    CombinationHelper.combinations(Enum.slice(list, (index-25)..(index-1)), 2)
      |> Enum.map(fn pair -> Enum.sum(pair) end)
  end
end

defmodule PartOne do
  def run(numbers) do
    sums = SumBuilder.build(numbers, 25)
    number = Enum.at(numbers, 25)
    if Enum.all?(sums, fn sum -> sum != number end) do
      number
    else
      [_ | tail] = numbers
      run(tail)
    end
  end
end

defmodule PartTwo do
  def run(numbers, target) do
    parts = Enum.reduce_while(numbers, [], fn n, acc ->
      if Enum.sum(acc) < target, do: {:cont, acc ++ [n] }, else: {:halt, acc}
    end)
    if Enum.sum(parts) == target do
      Enum.min(parts) + Enum.max(parts)
    else
      [_ | tail] = numbers
      run(tail, target)
    end
  end
end

part_one = PartOne.run(numbers)
IO.inspect(part_one)
PartTwo.run(numbers, part_one) |> IO.inspect
